File: python/src/kafka_tool/producer.py
import threading
from typing import Dict

from .data import ProducerConsumerData
from .producer_task import run_producer_task
from .reporter_task import run_reporter_task
from .settings import ProducerConsumerSettings
from .utils import recreate_topics, run_tasks


def run_producer(config: Dict[str, str], settings: ProducerConsumerSettings) -> None:
    if settings.recreate_topics: recreate_topics(config, settings)

    data = ProducerConsumerData()
    shutdown = threading.Event()
    threads = [
        threading.Thread(target=run_producer_task, args=[config, settings, data, producer_index, shutdown])
        for producer_index in range(settings.producers)]
    threads.append(threading.Thread(target=run_reporter_task, args=[config, data, shutdown]))

    run_tasks(threads, shutdown)

File: dotnet/KafkaTool/Producer.cs
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Spectre.Console.Cli;

namespace KafkaTool;

public class Producer : AsyncCommand<ProducerConsumerSettings>
{
    private static readonly ILogger Log = LoggerFactory
        .Create(builder => builder.AddSimpleConsole(options =>
        {
            options.SingleLine = true;
            options.TimestampFormat = "HH:mm:ss ";
        }))
        .CreateLogger("Log");
    
    public override async Task<int> ExecuteAsync(CommandContext context, ProducerConsumerSettings settings)
    {
        if (settings.RecreateTopics)
        {
            await Utils.RecreateTopics(settings);
        }

        var data = new ProducerConsumerData();
        
#if EXPERIMENTAL_ALLOC_FREE
        var producerTasks = Enumerable.Range(0, settings.Producers)
            .Select(producerIndex => ProducerAllocFreeTask.GetTask(settings, data, producerIndex));
#else
        var producerTasks = Enumerable.Range(0, settings.Producers)
            .Select(producerIndex => ProducerTask.GetTask(settings, data, producerIndex));
#endif
        
        var reporterTask = ReporterTask.GetTask(settings, data);
        var task = await Task.WhenAny(producerTasks.Concat([reporterTask]));
        await task;
        return 0;
    }
}
